### PR TITLE
WIP: fix RFC 1847 / Multipart/Signed

### PR DESCRIPTION
**Pull request by ahorek**
fixes https://github.com/mikel/mail/issues/1226

### CODE REVIEW

File: spec/mail/message_spec.rb [hunk 1]
```diff
@@ -1432,11 +1432,26 @@ def message_headers_should_match(message, other)
...
+      # the body of a "multipart/signed" entity.
+      it "rfc1847" do
+        mail = Mail.new
+        mail.body << Mail::Part.new.tap do |part|
+          part.content_disposition = 'attachment; filename="test.asc"'
+          part.content_type  = 'application/pgp-signature'
+        end
+        mail.encoded
+
+        expect(mail.multipart?).to eq(true)
```
Comment (jeremy) [inline spec/mail/message_spec.rb]: `be_truthy`

Comment (rafbm) [reply spec/mail/message_spec.rb]: @jeremy Out of curiosity, do you prefer `be_truthy` just for consistency across the project, a personal taste, or to leave room for `multipart?` returning a truthy object that isn’t `true`?

Comment (jeremy) [reply spec/mail/message_spec.rb]: All of the above, yes! In essence, follow Ruby's lead that implicit coercion to boolean is a desirable default. Use explicit boolean values when explicitly necessary, which is rare.

**Comment by skalee** [comment]
Thanks a lot, but no, it doesn't fix the issue.

The `application/pgp-signature` part may have any transfer encoding. The signature itself should be resilent to encoding changes — as are images, videos, or audios. Furthermore, signatures are typically in textual format. For instance, Open PGP encodes signature in it's own way called ASCII Armor (roughly base64, just different chars are used), and such signature can be transferred with any encoding, including 7bit, without any loss.

----

A multipart/signed message is structured as follows:

    container part or message (multipart/signed)
      |-> part which is signed (any mime type, incl. multipart)
      |-> signature (application/pgp-signature in case of Open PGP,
                     application/pkcs7-signature in case of S/MIME,
                     possibly others as well)

Content-Transfer-Encoding restrictions apply to the part which is signed, not to the signature.  It is required by standard that it's a first subpart of multipart/signed.  However, the MIME type of that part can be anything.

The reason why transfer encoding is important is that the signature is calculated over encoded message part.  The message composer first encodes the part which is about to be signed, then calculates a signature over that encoded part, and finally builds a multipart/signed.  One of the benefits of this behaviour is that it guarantees the integrity of MIME headers as well, esp. Content-Type and Content-ID.

Following example message comes from https://tools.ietf.org/html/rfc3156#section-5, and describes what happens:

      From: Michael Elkins <elkins@aero.org>
      To: Michael Elkins <elkins@aero.org>
      Mime-Version: 1.0

      Content-Type: multipart/signed; boundary=bar; micalg=pgp-md5;
        protocol="application/pgp-signature"

      --bar
    & Content-Type: text/plain; charset=iso-8859-1
    & Content-Transfer-Encoding: quoted-printable
    &
    & =A1Hola!
    &
    & Did you know that talking to yourself is a sign of senility?
    &
    & It's generally a good idea to encode lines that begin with
    & From=20because some mail transport agents will insert a greater-
    & than (>) sign, thus invalidating the signature.
    &
    & Also, in some cases it might be desirable to encode any   =20
    & trailing whitespace that occurs on lines in order to ensure  =20
    & that the message signature is not invalidated when passing =20
    & a gateway that modifies such whitespace (like BITNET). =20
    &
    & me

      --bar

      Content-Type: application/pgp-signature

      -----BEGIN PGP MESSAGE-----
      Version: 2.6.2

      iQCVAwUBMJrRF2N9oWBghPDJAQE9UQQAtl7LuRVndBjrk4EqYBIb3h5QXIX/LC//
      jJV5bNvkZIGPIcEmI5iFd9boEgvpirHtIREEqLQRkYNoBActFBZmh9GC3C041WGq
      uMbrbxc+nIs1TIKlA08rVi9ig/2Yh7LFrK5Ein57U/W72vgSxLhe/zhdfolT9Brn
      HOxEa44b+EI=
      =ndaj
      -----END PGP MESSAGE-----

      --bar--

With following note:

> The "&"s in the previous example indicate the portion of the data over which the signature was calculated.

If some gateway would transform the signed part to base64 or quoted-printable, it would produce a different sequence of bytes, making signature invalid.  Parts encoded as quoted-printable or base64 are never altered during transport, hence are safe to use.

----

BTW I need to check what transfer encoding should be used when signing a multipart part.

**Comment by ahorek** [comment]
thanks for explanation! let me will fix it

**Pull request closed without merging**